### PR TITLE
Fix Poste email validation in Rails 4.2;  add tests.

### PR DESCRIPTION
**Pull request by philbogle**
Updates the Poste2.email_address? method to be compatible with Rails 4.2;   adds tests.

The original implementation was fragile because it called a private method in the implementation of Mail::Address that is not available in later versions, and because it blindly catches all exceptions (including NoMethodError!) causing it spuriously return false on valid email address in Rails 4.2 rather than giving a helpful exception. Also, there were no tests.

I reworked the implementation to not rely on private implementation details and added test to verify expect behavior.  (Note that there are some technically invalid email address that the original implementation still accept, like ".name@example.com".  For now I thought it was OK to preserve existing behavior.)

I had to refactor the email_address? method into a separate module for testability because Poste is impossible to test in its current form.  It relies on db.rb which attempts to open up the poste database upon being required, using configuration values that are not available in test environment.

### CODE REVIEW

File: shared/test/test_email_validator.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+      'email@example.museum',
+      'email@example.co.jp',
+      'firstname-lastname@example.com']
+
+    valid_addresses.each do |address|
+      assert EmailValidator::email_address?(address), "#{address} should be a valid email address"
+    end
+  end
+
+  def test_invalid_addresses
```
Comment (laurelfan) [inline shared/test/test_email_validator.rb]: what is the behavior for nil and empty string? Things that are not strings?

File: lib/cdo/email_validator.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+    return false unless email.address == address  # Must be well-formed
+    return false unless email.domain  # Must have a domain
+
+    # Reject single part domains like "localhost".
+    domain_parts = email.domain.split('.')
+    return false if domain_parts.length < 2
+
+    # Reject invalid domains like "example..com"
+    return false if domain_parts.detect {|p| p == ""}
```
Comment (laurelfan) [inline lib/cdo/email_validator.rb]: `domain_parts.include?("")` would work too

Comment (philbogle) [inline lib/cdo/email_validator.rb]: I like that better, done.

File: shared/test/test_email_validator.rb (hunk 1 above)
Comment (philbogle) [inline shared/test/test_email_validator.rb]: Done. Added checks for nil, empty string, numbers, and hashes, none of which should be considered valid email addresses.

**Comment by philbogle** [comment]
Responded to Laurel's comments. Thank you for the review!

**Comment by laurelfan** [comment]
lgtm :+1:

**Pull request merged**